v3/src/components/common/formula-insert-values-menu.tsx

```diff
@@ -0,0 +1,190 @@
+  }
+
+  useEffect(() => {
+    setListContainerStyle(getListContainerStyle())
+
+    const handleScrollPosition = () => {
+      if (scrollableContainerRef.current) {
+        setScrollPosition(scrollableContainerRef.current.scrollTop)
+      }
+    }
+
+    const container = scrollableContainerRef.current
+    container?.addEventListener("scroll", handleScrollPosition)
+
+    return () => {
+      container?.removeEventListener("scroll", handleScrollPosition)
+    }
+  }, [])
+
+  const isScrollable = scrollableContainerRef.current && scrollableContainerRef.current.scrollHeight > kMaxHeight
+  const canScrollUp = scrollableContainerRef.current && scrollableContainerRef.current.scrollTop > 0
+  const canScrollDown = scrollableContainerRef.current &&
+          (scrollableContainerRef.current.scrollHeight - scrollableContainerRef.current.scrollTop + 20 > kMaxHeight)
+
+  return (
+    <Flex className="formula-operand-list-container" data-testid="formula-value-list"
+        style={listContainerStyle} >
+      { isScrollable && canScrollUp &&
+      <div className="scroll-arrow" onPointerOver={()=>handleScroll("up")}>
+        <TriangleUpIcon />
+      </div>
+      }
+      <List className="formula-operand-scrollable-container"  ref={scrollableContainerRef}>
+        <List className="formula-operand-list">
+          { collections?.map((collection, index) => {
+            return (
+              <React.Fragment key={collection.id}>
+                <List className="formula-operand-subset">
+                  { attributeNamesInCollection?.[index]?.map((attrName) => {
+                    return (
+                      <ListItem className="formula-operand-list-item" key={attrName}
+                          onClick={() => insertValueToFormula(attrName)} data-testid="formula-value-item">
+                        {attrName}
+                      </ListItem>
+                    )
+                  })}
+                </List>
+                <Divider className="list-divider"/>
+              </React.Fragment>
+            )
+          })}
+        </List>
+        <List className="formula-operand-subset">
+          <ListItem className="formula-operand-list-item" onClick={() => insertValueToFormula("caseIndex")}>
+            caseIndex
+          </ListItem>
+        </List>
+        <Divider className="list-divider"/>
+        <List className="formula-operand-subset">
+          { remoteBoundaryData.map((boundary) => {
+            return (
+              <ListItem key={boundary} className="formula-operand-list-item"
+                    onClick={() => insertValueToFormula(boundary)}>
+                {boundary}
+              </ListItem>
+            )
+          })}
+          { globals.map((global) => {
+            return (
+              <ListItem key={global.label} className="formula-operand-list-item"
+                    onClick={() => insertValueToFormula(global.label)}>
+                {global.label}
+              </ListItem>
```

Comment:
This pattern is repeated five times in this component. It would be better to make a separate `InsertValuesMenuListItem` component rather than duplicating the same code this much.